### PR TITLE
Diag domain config

### PR DESCRIPTION
**Pull request by florian-sanders-cc**
## What does this PR do?

- Moves types & utils related to domains from components to the client project,
- Adds unit tests for the domain utils,
- Adds a `diagDomainConfig` helper + associated tests.

Some important context:

- we throw errors and tell users their DNS config is wrong when:
  - no config was found (no CNAME / no A records),
  - the config contains at least one unknown record (a record that does not match the default Load Balancer config of the app).
- APEX domains can only be configured with A records so diagnostics for such domains will never mention CNAME records,
- we don't support configs relying on a CDN (like Cloudflare for instance) and we don't support private Load Balancers for now (no API).

### CODE REVIEW

File: esm/utils/diag-domain-config.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    diagSummary: getSummaryCode(diagDetails),
+    diagDetails,
+  };
+}
+
+/**
+   * @param {Array<string>} aRecords
+   * @param {LoadBalancerDnsConfig} loadBalancerDnsConfig
+   * @returns {Array<RecordDiag>}
+   */
```
Comment (hsablonniere) [inline esm/utils/diag-domain-config.js:88]: nitpick: small indentation pb

File: test/diag-domain-config.spec.js [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+
+      expect(domainDiagResults).to.deep.equal({
+        hostname: domainInfo.hostname,
+        pathPrefix: domainInfo.pathPrefix,
+        isApex: domainInfo.isApex,
+        isWildcard: domainInfo.isWildcard,
+        diagSummary: 'managed',
+        diagDetails: [],
+      });
+    });
```
Comment (hsablonniere) [inline test/diag-domain-config.spec.js:846]: issue: this should be a `it` instead of a `describe`

**Review by hsablonniere — CHANGES_REQUESTED** [review]
Impressive work, well done!!! :clap: 

Praise for the typechecking :heart_eyes:

File: esm/utils/domains.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      isWildcard,
+    };
+  }
+  catch {
+    if (domain.length === 0) {
+      throw new DomainParseError('empty', 'Empty domain value');
+    }
+
+    throw new DomainParseError('invalid-format', 'Invalid domain format');
+  }
```
Comment (pdesoyres-cc) [inline esm/utils/domains.js:36]: Here you loose the cause of the error.

```suggestion
  catch (e) {
    if (domain.length === 0) {
      throw new DomainParseError('empty', 'Empty domain value');
    }

    throw new DomainParseError('invalid-format', 'Invalid domain format', e);
  }
```

File: esm/utils/domains.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+export class DomainParseError extends Error {
+  /**
+   * @param {'empty'|'invalid-wildcard'|'invalid-format'} code
+   * @param {string} message
+   */
+  constructor (code, message) {
+    super(message);
+
+    this.code = code;
+  }
```
Comment (pdesoyres-cc) [inline esm/utils/domains.js:48]: Add an optional `cause` parameter:

```suggestion
  constructor (code, message, cause) {
    super(message);

    this.code = code;
    this.cause = cause;
  }
```

File: esm/utils/diag-domain-config.types.d.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+export interface RecordDiag {
+  code:
+    | "valid-a"
+    | "unknown-a"
+    | "missing-a"
+    | "suggested-cname"
+    | "unknown-cname"
+    | "missing-cname"
+    | "test-only"; // Union type 'valid-cname'|'missing-cname'|'
```
Comment (pdesoyres-cc) [inline esm/utils/diag-domain-config.types.d.ts:44]: question: is this comment useful?

**Review by pdesoyres-cc — APPROVED** [review]
LGTM. I have just some minor suggestions.

Well Done Florian.

File: esm/utils/diag-domain-config.types.d.ts (hunk 5 above)
Comment (florian-sanders-cc) [reply esm/utils/diag-domain-config.types.d.ts:44]: nope, thanks for pointing it out :sweat_smile:

**Pull request merged**